core/store/src/genesis.rs

```diff
@@ -0,0 +1,147 @@
+    let shard_layout = initial_epoch_config.shard_layout;
+    let num_shards = shard_layout.num_shards();
+    let mut shard_account_ids: Vec<HashSet<AccountId>> =
+        (0..num_shards).map(|_| HashSet::new()).collect();
+    let mut has_protocol_account = false;
+    info!(target: "store","distributing records to shards");
+
+    genesis.for_each_record(|record: &StateRecord| {
+        shard_account_ids[state_record_to_shard_id(record, &shard_layout) as usize]
+            .insert(state_record_to_account_id(record).clone());
+        if let StateRecord::Account { account_id, .. } = record {
+            if account_id == &genesis.config.protocol_treasury_account {
+                has_protocol_account = true;
+            }
+        }
+    });
+    assert!(has_protocol_account, "Genesis spec doesn't have protocol treasury account");
+    let tries = ShardTries::new(
+        store.clone(),
+        TrieConfig::default(),
+        &genesis.config.shard_layout.get_shard_uids(),
+        FlatStorageManager::new(store),
+    );
+
+    let runtime_config_store = create_runtime_config_store(&genesis.config.chain_id);
+    let runtime_config = runtime_config_store.get_config(genesis.config.protocol_version);
+    let writers = std::sync::atomic::AtomicUsize::new(0);
+    (0..num_shards)
+        .into_par_iter()
+        .map(|shard_id| {
+            let validators = genesis
+                .config
+                .validators
+                .iter()
+                .filter_map(|account_info| {
+                    if account_id_to_shard_id(&account_info.account_id, &shard_layout) == shard_id {
+                        Some((
+                            account_info.account_id.clone(),
+                            account_info.public_key.clone(),
+                            account_info.amount,
+                        ))
+                    } else {
+                        None
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            GenesisStateApplier::apply(
+                &writers,
+                tries.clone(),
+                shard_id,
+                &validators,
+                runtime_config,
+                genesis,
+                shard_account_ids[shard_id as usize].clone(),
+            )
+        })
+        .collect()
+}
+
+fn state_record_to_shard_id(state_record: &StateRecord, shard_layout: &ShardLayout) -> ShardId {
+    account_id_to_shard_id(state_record_to_account_id(state_record), shard_layout)
+}
+
+/// Create store of runtime configs for the given chain id.
+///
+/// For mainnet and other chains except testnet we don't need to override runtime config for
+/// first protocol versions.
+/// For testnet, runtime config for genesis block was (incorrectly) different, that's why we
+/// need to override it specifically to preserve compatibility.
+pub fn create_runtime_config_store(chain_id: &str) -> RuntimeConfigStore {
```

Comment:
This is not a big issue for now, but I want to avoid creating dependencies of store on runtime. It should be the opposite - runtime uses store engine to read state data, not the vice versa.

I've looked into the code - it seems that we need RuntimeConfigStore just for `StorageComputer::new(config)` which itself uses only StorageUsageConfig. Can we move `create_runtime_config_store` back to NightshadeRuntime and refactor code so that genesis logic here needs only StorageUsageConfig?